pkg/ccl/changefeedccl/changefeedbase/target.go

```diff
@@ -25,8 +25,102 @@ type Target struct {
 // the changefeed, possibly modified by WITH options.
 type StatementTimeName string
 
-// WatchedTables tracks the StatementTimeName for each table id.
-type WatchedTables map[descpb.ID]StatementTimeName
+type targetsByTable struct {
+	wholeTable   *Target
+	byFamilyName map[string]Target
+}
+
+func (tbt targetsByTable) add(t Target) targetsByTable {
+	if t.FamilyName == "" {
+		tbt.wholeTable = &t
+	} else {
+		if tbt.byFamilyName == nil {
+			tbt.byFamilyName = make(map[string]Target)
+		}
+		tbt.byFamilyName[t.FamilyName] = t
+	}
+	return tbt
+}
+
+func (tbt targetsByTable) each(f func(t Target) error) error {
+	if tbt.wholeTable != nil {
+		if err := f(*tbt.wholeTable); err != nil {
+			return err
+		}
+		for _, t := range tbt.byFamilyName {
+			if err := f(t); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
 
 // Targets is the complete list of target specifications for a changefeed.
-type Targets []Target
+// This is stored as a map of TableID -> Family Name -> Target in order
+// to support all current ways we need to iterate over it.
+type Targets struct {
+	Size uint
```

Comment:
Because of how it's nested, I need to store the size anyway if I want getting it to be constant time, so I figured might as well make it public. Can make it a getter instead if you think that's better.